crates/ra_ide/src/references.rs

```diff
@@ -58,24 +74,43 @@ impl ReferenceSearchResult {
 // allow turning ReferenceSearchResult into an iterator
 // over FileRanges
 impl IntoIterator for ReferenceSearchResult {
-    type Item = FileRange;
-    type IntoIter = std::vec::IntoIter<FileRange>;
+    type Item = Reference;
+    type IntoIter = std::vec::IntoIter<Reference>;
 
     fn into_iter(mut self) -> Self::IntoIter {
         let mut v = Vec::with_capacity(self.len());
-        v.push(FileRange { file_id: self.declaration.file_id(), range: self.declaration.range() });
+        v.push(Reference {
+            file_range: FileRange {
+                file_id: self.declaration.file_id(),
+                range: self.declaration.range(),
+            },
+            kind: self.declaration_kind,
+        });
         v.append(&mut self.references);
         v.into_iter()
     }
 }
 
 pub(crate) fn find_all_refs(
     db: &RootDatabase,
-    position: FilePosition,
+    mut position: FilePosition,
     search_scope: Option<SearchScope>,
 ) -> Option<RangeInfo<ReferenceSearchResult>> {
     let parse = db.parse(position.file_id);
     let syntax = parse.tree().syntax().clone();
+
+    let token = syntax.token_at_offset(position.offset);
+    let mut search_kind = ReferenceKind::Other;
+
+    if let TokenAtOffset::Between(ref left, ref right) = token {
+        if (right.kind() == SyntaxKind::L_CURLY || right.kind() == SyntaxKind::L_PAREN)
+            && left.kind() != SyntaxKind::IDENT
```

Comment:
I added `left.kind() != SyntaxKind::IDENT` to ignore `fn_name()` case but now I think It should be something similar to look for `RecordLit` as I added below and add support for tuples.